File: src/algorithms/trees/bst-operations/bst-range-sum-iterative/sources/bst-range-sum-iterative_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("bst-range-sum-iterative")
BSTNode = module.BSTNode
bst_range_sum_iterative = module.bst_range_sum_iterative


def make_node(value, left=None, right=None):
    node = BSTNode(value)
    node.left = left
    node.right = right
    return node


tree = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))


def test_sums_range_3_to_7():
    assert bst_range_sum_iterative(tree, 3, 7) == 3 + 4 + 5 + 6 + 7


def test_sums_all_values():
    assert bst_range_sum_iterative(tree, 1, 7) == 28


def test_returns_zero_no_match():
    assert bst_range_sum_iterative(tree, 10, 20) == 0


def test_null_tree():
    assert bst_range_sum_iterative(None, 1, 7) == 0


if __name__ == "__main__":
    test_sums_range_3_to_7()
    test_sums_all_values()
    test_returns_zero_no_match()
    test_null_tree()
    print("All tests passed!")

File: src/algorithms/trees/bst-operations/bst-range-sum-iterative/sources/bst-range-sum-iterative_test.rs
include!("bst-range-sum-iterative.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BSTNode>>, right: Option<Box<BSTNode>>) -> Option<Box<BSTNode>> {
        Some(Box::new(BSTNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BSTNode>> {
        make_node(value, None, None)
    }

    fn build_tree() -> Option<Box<BSTNode>> {
        make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)),
        )
    }

    #[test]
    fn test_sums_range_3_to_7() {
        assert_eq!(bst_range_sum_iterative(&build_tree(), 3, 7), 25);
    }

    #[test]
    fn test_sums_all_values() {
        assert_eq!(bst_range_sum_iterative(&build_tree(), 1, 7), 28);
    }

    #[test]
    fn test_no_match_returns_zero() {
        assert_eq!(bst_range_sum_iterative(&build_tree(), 10, 20), 0);
    }

    #[test]
    fn test_null_tree() {
        assert_eq!(bst_range_sum_iterative(&None, 1, 7), 0);
    }
}
